photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/LBFGS.scala

```diff
@@ -15,126 +15,99 @@
-   * Breeze to optimize functions with L1 penalty term. The diffFunction is also translated to a form these breeze
-   * optimizer can understand.
-   * The L1 penalty is implemented in the optimizer level. See
-   * [[http://www.scalanlp.org/api/breeze/index.html#breeze.optimize.OWLQN breeze.optimize.OWLQN]].
-   */
-  protected[ml] class BreezeOptimization(
-      data: Either[RDD[Datum], Iterable[Datum]],
-      diffFunction: DiffFunction[Datum],
-      initialCoef: Vector[Double]) {
+  * Class used to solve an optimization problem using Limited-memory BFGS (LBFGS).
+  * Reference: [[http://en.wikipedia.org/wiki/Limited-memory_BFGS]].
+  *
+  * @param numCorrections The number of corrections used in the LBFGS update. Default 10. Values of numCorrections less
+  *                       than 3 are not recommended; large values of numCorrections will result in excessive computing
+  *                       time.
+  *                       Recommended:  3 < numCorrections < 10
+  *                       Restriction:  numCorrections > 0
+  * @param tolerance The tolerance threshold for improvement between iterations as a percentage of the initial loss
+  * @param maxNumIterations The cut-off for number of optimization iterations to perform.
+  * @param constraintMap (Optional) The map of constraints on the feature coefficients
+  * @param isTrackingState Whether to track intermediate states during optimization
+  * @param isReusingPreviousInitialState Whether to reuse the previous initial state or not. When warm-start training is
+  *                                      desired, i.e. in grid-search based hyper-parameter tuning, this field is
+  *                                      recommended to set to true for consistent convergence check.
+  */
+class LBFGS(
+    numCorrections: Int = LBFGS.DEFAULT_NUM_CORRECTIONS,
+    tolerance: Double = LBFGS.DEFAULT_TOLERANCE,
+    maxNumIterations: Int = LBFGS.DEFAULT_MAX_ITER,
+    constraintMap: Option[Map[Int, (Double, Double)]] = Optimizer.DEFAULT_CONSTRAINT_MAP,
+    isTrackingState: Boolean = Optimizer.DEFAULT_TRACKING_STATE,
+    isReusingPreviousInitialState: Boolean = Optimizer.DEFAULT_REUSE_PREVIOUS_INIT_STATE)
+  extends Optimizer[DiffFunction](
+    tolerance,
+    maxNumIterations,
+    constraintMap,
+    isTrackingState,
+    isReusingPreviousInitialState) {
 
-    // Initialize the solver (L-BFGS for L2 and OWLQN for L1/Elastic net)
-    private val lbfgs: BreezeLBFGS[Vector[Double]] = diffFunction match {
-      case diffFunc: DiffFunction[Datum] with L1RegularizationTerm =>
-        val l1Weight = diffFunc.getL1RegularizationParam
-        new BreezeOWLQN[Int, Vector[Double]](maxNumIterations, numCorrections, (_: Int) => l1Weight, tolerance)
-      case diffFunc: DiffFunction[Datum] =>
-        new BreezeLBFGS[Vector[Double]](maxNumIterations, numCorrections, tolerance)
-    }
+  protected val breezeOptimizer = new BreezeLBFGS[Vector[Double]](maxNumIterations, numCorrections, tolerance)
```

Comment:
@joshvfleming 
OWLQN has become its own optimizer. The selection between LBFGS and OWLQN is now made within the OptimizerFactory.